// boost/lib/Command.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _Pipeline = require('./Pipeline');

var _Pipeline2 = _interopRequireDefault(_Pipeline);

var _Routine = require('./Routine');

var _Routine2 = _interopRequireDefault(_Routine);

var _Tool = require('./Tool');

var _Tool2 = _interopRequireDefault(_Tool);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class Command {

  constructor(appName, command) {
    this.routines = [];

    this.appName = appName;
    this.command = command;

    command.action(args => this.run(args));
  }

  alias(...aliases) {
    this.command.alias(...aliases);

    return this;
  }

  autocomplete(list) {
    this.command.autocomplete(list);

    return this;
  }

  option(...args) {
    this.command.option(...args);

    return this;
  }

  pipe(...routines) {
    this.routines.push(...routines);

    return this;
  }

  run(value) {
    const context = {};

    const tool = new _Tool2.default(this.appName).loadConfig().loadPlugins();

    return new _Pipeline2.default(tool).pipe(...this.routines).run(value, context);
  }

  types(types) {
    this.command.types(types);

    return this;
  }

  validate(callback) {
    this.command.validate(callback);

    return this;
  }
}
exports.default = Command; /**
                            * @copyright   2017, Miles Johnson
                            * @license     https://opensource.org/licenses/MIT
                            * 
                            */
